pgbouncer/datadog_checks/pgbouncer/pgbouncer.py

```diff
@@ -87,51 +95,41 @@ def _collect_stats(self, db, instance_tags):
         metric_scope = [self.STATS_METRICS, self.POOLS_METRICS]
 
         try:
-            cursor = db.cursor()
-            results = None
-
-            for scope in metric_scope:
-
-                metrics = scope['metrics']
-                cols = [m[0] for m in metrics]
-
-                try:
+            with db.cursor(cursor_factory=pgextras.DictCursor) as cursor:
+                for scope in metric_scope:
+                    descriptors = scope['descriptors']
+                    metrics = scope['metrics']
                     query = scope['query']
-                    self.log.debug("Running query: %s" % query)
-                    cursor.execute(query)
 
-                    results = cursor.fetchall()
+                    try:
+                        self.log.debug("Running query: %s", query)
+                        cursor.execute(query)
 
-                except pg.Error as e:
-                    self.log.warning("Not all metrics may be available: %s" % str(e))
-                    continue
+                        rows = cursor.fetchall()
 
-                for row in results:
-                    if row[0] == self.DB_NAME:
-                        continue
+                    except pg.Error as e:
+                        self.log.exception("Not all metrics may be available")
 
-                    desc = scope['descriptors']
+                    else:
+                        for row in rows:
+                            self.log.debug("Processing row: %r", row)
 
-                    # Some versions of pgbouncer have extra fields at the end of SHOW POOLS
-                    if len(row) == len(cols) + len(desc) + 1:
-                        row = row[:-1]
-                    elif len(row) == len(cols) + len(desc) + 2:
-                        row = row[:-2]
+                            # Skip the "pgbouncer" database
+                            if row['database'] == self.DB_NAME:
+                                continue
 
-                    assert len(row) == len(cols) + len(desc)
+                            tags = list(instance_tags)
+                            tags += ["%s:%s" % (tag, row[column]) for (column, tag) in descriptors]
```

Comment:
We should test if `row` contains `column` to avoid raising an exception if the column in `descriptor` does not exist.
Something like:
```python
 tags += ["%s:%s" % (tag, row[column]) for (column, tag) in descriptors if column in row]
```

Reply:
These are pretty important and should never be missing. If they are missing then it should probably explode.

Reply:
I'll add the conditional anyway as protection against column name changes :+1: